site/sfguides/src/production_ready_rag_with_trulens_snowflake_cortex/production_ready_rag_trulens_snowflake_cortex.md

```diff
@@ -0,0 +1,483 @@
+  name="Context Relevance")
+  .on_input()
+  .on(Select.RecordCalls.retrieve_context.rets[:])
+  .aggregate(np.mean)
+)
+
+f_answer_relevance = (
+  Feedback(
+  provider.relevance,
+  name="Answer Relevance")
+  .on_input()
+  .on_output()
+  .aggregate(np.mean)
+)
+
+feedbacks = [f_context_relevance,
+      f_answer_relevance,
+      f_groundedness,
+    ]
+```
+
+```python
+from trulens_eval import TruCustomApp
+tru_rag = TruCustomApp(rag,
+  app_id = 'RAG v1',
+  feedbacks = [f_groundedness, f_answer_relevance, f_context_relevance])
+```
+
+## Test the application and observe performance
+Duration: 3
+
+```python
+prompts = [
+  "How do I launch a streamlit app?",
+  "How can I capture the state of my session in streamlit?",
+  "How do I install streamlit?",
+  "How do I change the background color of a streamlit app?",
+  "What's the advantage of using a streamlit form?",
+  "What are some ways I should use checkboxes?",
+  "How can I conserve space and hide away content?",
+  "Can you recommend some resources for learning Streamlit?",
+  "What are some common use cases for Streamlit?",
+  "How can I deploy a streamlit app on the cloud?",
+  "How do I add a logo to streamlit?",
+  "What is the best way to deploy a Streamlit app?",
+  "How should I use a streamlit toggle?",
+  "How do I add new pages to my streamlit app?",
+  "How do I write a dataframe to display in my dashboard?",
+  "Can I plot a map in streamlit? If so, how?",
+  "How do vector stores enable efficient similarity search?",
+  "How do I prevent my child from using the internet?",
+  "What should I pack for a camping trip this weekend?",
+  "How do I defend myself against bear attacks?"
+]
+```
+
+```python
+with tru_rag as recording:
+  for prompt in prompts:
+    rag.query(prompt)
+
+tru.get_leaderboard()
+```
+
+## Use Guardrails
+Duration: 7
+
+In addition to making informed iteration, we can also directly use feedback results as guardrails at inference time. In particular, here we show how to use the context relevance score as a guardrail to filter out irrelevant context before it gets passed to the LLM. This both reduces hallucination and improves efficiency.
+
+![Context Filter Guardrails](./assets/guardrail_context_filtering.png)
+
+To do so, we'll rebuild our RAG using the `@context-filter decorator` on the method we want to filter, and pass in the feedback function and threshold to use for guardrailing.
+
+```python
+# note: feedback function used for guardrail must only return a score, not also reasons
+f_context_relevance_score = (
+  Feedback(provider.context_relevance, name = "Context Relevance")
+)
+
+from trulens_eval.guardrails.base import context_filter
+
+# note: feedback function used for guardrail must only return a score, not also reasons
+f_context_relevance_score = (
+  Feedback(provider.context_relevance, name = "Context Relevance")
+  .on_input()
+  .on(Select.RecordCalls.retrieve.rets)
+)
+
+from trulens_eval.guardrails.base import context_filter
```

Comment:
you import this twice (in L431 before). Also in general imports come at the top anyway.